aten/src/ATen/native/QuantizedLinear.cpp

```diff
@@ -0,0 +1,192 @@
+    fbgemm::matrix_op_t::Transpose,
+    K,
+    1,
+    static_cast<int8_t*>(input.data_ptr()),
+    K,
+    nullptr,
+    1);
+
+  PtrPun pun;
+  pun.ptr = ptr;
+
+  auto retval = input.clone().resize_({1}).to(at::kLong);
+  auto retval_accessor = retval.accessor<int64_t, 1>();
+  retval_accessor[0] = pun.integral;
+
+  return retval;
+}
+
+Tensor linear_int8_weight(const Tensor& input, const Tensor& weight, const Tensor& packed, Scalar weight_scale, Scalar weight_zero_point, const Tensor& bias) {
+  // Dynamic dispatch for machines that don't support AVX2
+  // Note that we assume `weight` is unquantized and scale and zero_point
+  // are dont-care
+  if (!fbgemm::fbgemmSupportedCPU()) {
+    return at::native::linear(input, weight, bias);
+  }
+
+  auto *input_ptr = input.data<float>();
+  auto *weight_ptr = (int8_t*)weight.data<uint8_t>();
+
+  AT_ASSERT(input.dim() >= 2);
+  int64_t M = 1;
+  for (size_t i = 0; i < input.dim() - 1; ++i)  {
+    M *= input.size(i);
+  }
+  int64_t K = input.size(input.dim() - 1);
+  AT_ASSERT(weight.dim() == 2);
+  AT_ASSERT(K == weight.size(1));
+  auto N = weight.size(0);
+  AT_ASSERT(bias.dim() == 1);
+  AT_ASSERT(bias.size(0) == N);
+  AT_ASSERT(weight_scale.isFloatingPoint());
+  AT_ASSERT(weight_zero_point.isIntegral());
+
+  float x_min, x_max;
+  fbgemm::FindMinMax(input_ptr, &x_min, &x_max, input.numel());
+
+  static constexpr int precision = 8;
+  static constexpr bool is_signed = false;
+
+  auto q_params = fbgemm::ChooseQuantizationParams(
+    x_min,
+    x_max,
+    is_signed ? -(1 << (precision - 1)) : 0,
+    is_signed ? ((1 << (precision - 1)) - 1) : (1 << precision) - 1,
+    false);
+
+  q_params.precision = precision;
+
+  using namespace fbgemm;
+
+  std::vector<int32_t> row_offsets_, column_offsets_;
+  column_offsets_.resize(M * N);
+  row_offsets_.resize(
+      PackAWithQuantRowOffset<uint8_t>::rowOffsetBufferSize());
+  static std::vector<uint8_t> X_pack_buf_;
+  X_pack_buf_.resize(
+      PackAWithQuantRowOffset<uint8_t>::packedBufferSize());
+  PackAWithQuantRowOffset<uint8_t> packA(
+      matrix_op_t::NoTranspose,
+      M,
+      K,
+      input_ptr,
+      K,
+      X_pack_buf_.data(), // buffer for packed matrix
+      q_params.scale,
+      q_params.zero_point,
+      1, // groups
+      row_offsets_.data());
+
+  float weight_scale_float = static_cast<float>(weight_scale.to<double>());
+  int32_t weight_zero_point_int32 = static_cast<int32_t>(weight_zero_point.to<int64_t>());
+
+  DoNothing<float, float> doNothingObj{};
+  fbgemm::ReQuantizeForFloat<false /* FUSE_RELU*/> outputProcObj(
+      doNothingObj,
+      q_params.scale,
+      &weight_scale_float,
+      q_params.zero_point,
+      &weight_zero_point_int32,
+      packA.getRowOffsetBuffer(),
+      column_offsets_.data(),
+      bias.data<float>(),
+      /*ncol=*/N); // bias
+
+  auto output = bias.to(at::kFloat).expand({M, N}).contiguous();
+
+  auto packed_ptr_accessor = packed.accessor<int64_t, 1>();
```

Comment:
can't it be just packed.item()?

also - define a subfunction that does pointer mangling inside, like GetResourcePointerPun(tensor)